python/neuroglancer/ingest/volumes/gcloudvolume.py

```diff
@@ -0,0 +1,199 @@
+import os
+import json
+import re
+
+import numpy as np
+from tqdm import tqdm
+
+import neuroglancer
+
+from volumes import Volume
+from google.cloud import storage as gstorage
+
+import lib
+from lib import clamp, mkdir, list_shape, xyzrange, Vec3, Bbox, map2, min2, max2
+from lib import GCLOUD_PROJECT_NAME, COMMON_STAGING_DIR
+
+class GCloudVolume(Volume):
+  def __init__(self, dataset_name, layer, mip):
+    super(self.__class__, self).__init__()
+
+    self.dataset_name = dataset_name
+    self.layer = layer
+    self.mip = mip
+    
+    self.cacheFiles = True
+
+    blob = self.__getInfoBlob()
+
+    self.info = json.loads(blob.download_as_string())
+
+  def __getInfoBlob(self):
+    client = gstorage.Client(project=lib.GCLOUD_PROJECT_NAME)
+    bucket = client.get_bucket(lib.GCLOUD_BUCKET_NAME)
+
+    info_path = os.path.join(self.dataset_name, self.layer, 'info')
+    return gstorage.blob.Blob(info_path, bucket)
+
+  def addScale(self, factor):
+    # e.g. {"encoding": "raw", "chunk_sizes": [[64, 64, 64]], "key": "4_4_40", 
+    # "resolution": [4, 4, 40], "voxel_offset": [0, 0, 0], 
+    # "size": [2048, 2048, 256]}
+    fullres = self.info['scales'][0]
+
+    newscale = {
+      u"encoding": fullres['encoding'],
+      u"chunk_sizes": fullres['chunk_sizes'],
+      u"resolution": list( Vec3(*fullres['resolution']) * factor ),
+      u"voxel_offset": fullres['voxel_offset'],
+      u"size": list(np.ceil(Vec3(*fullres['size']) / Vec3(*factor)).astype(int)),
+    }
+
+    newscale[u'key'] = unicode("_".join([ str(res) for res in newscale['resolution']]))
+
+    new_res = np.array(newscale['resolution'], dtype=int).astype(int)
+
+    preexisting = False
+    for index, scale in enumerate(self.info['scales']):
+      res = np.array(scale['resolution'], dtype=int).astype(int)
+      if np.array_equal(new_res, res):
+        preexisting = True
+        self.info['scales'][index] = newscale
+        break
+
+    if not preexisting:    
+      self.info['scales'].append(newscale)
+
+    return newscale
+
+  def commit(self):
```

Comment:
It gets the current blob. I don't save a copy of the blob reference, just the data in self.info. This uploads the current version of info.